File: 04-distributed/03-clocks-lie/golang/clock_audit.go
// Layer 4 Topic 3 (Part A) -- Go's clocks, audited rather than assumed.
//
// WHAT THIS DEMONSTRATES: four things, in order.
//  1. the clock inventory. Go hides the distinction on purpose: a time.Time
//     carries BOTH a wall reading and a monotonic reading, and t2.Sub(t1)
//     silently uses the monotonic one. Idiomatic Go is correct here without the
//     author knowing why -- which is exactly what makes item 3 dangerous.
//  2. one span timed twice, through the application's own now() (wall) and
//     through the monotonic reading, with an NTP-style step inside two spans.
//  3. the footgun specific to this runtime: the monotonic reading is STRIPPED by
//     t.Round, t.Truncate, t.UTC, t.Local, t.AddDate, JSON marshal/unmarshal and
//     anything from time.Parse. A duration computed after any of those falls back
//     to the wall clock silently and can come out negative. This program shows
//     each stripper, and then computes the same span both ways so the two paths
//     visibly diverge.
//  4. the summary line for the README's record table.
//
// WHAT TO LOOK FOR IN THE OUTPUT: section 3's table of strippers -- the "m=+..."
// suffix in a printed time.Time is the monotonic reading, and its disappearance
// is the whole bug. Then the last two lines of section 3, where the same span
// measured through a round-tripped Time comes out NEGATIVE while the untouched
// one is correct.
//
//	cd golang && go run clock_audit.go
package main

import (
	"encoding/json"
	"fmt"
	"runtime"
	"sort"
	"time"
)

const (
	stepBack  = -40 * time.Second // an NTP correction, applied backwards, mid-run
	spans     = 400
	spanWorkU = 200 * time.Microsecond
)

// ------------------------------------------------------------- 1. inventory

// measureResolution returns the smallest non-zero delta this clock reports.
// Measured, not documented: a clock can advertise nanoseconds and tick in
// microseconds, and on Darwin several of them do.
func measureResolution(read func() int64, trials int) int64 {
	smallest := int64(1 << 62)
	for i := 0; i < trials; i++ {
		a := read()
		for {
			b := read()
			if b != a {
				d := b - a
				if d < 0 {
					d = -d
				}
				if d < smallest {
					smallest = d
				}
				break
			}
		}
	}
	return smallest
}

func inventory() {
	fmt.Println("------------------------------------------------------------------------------")
	fmt.Println("1. the clocks Go offers, and the one it hides")
	fmt.Println("------------------------------------------------------------------------------")
	fmt.Printf("  %-40s%-12s%s\n", "expression", "kind", "measured resolution")

	start := time.Now()
	rows := []struct {
		name string
		kind string
		read func() int64
	}{
		{"time.Now().UnixNano()", "realtime", func() int64 { return time.Now().UnixNano() }},
		{"time.Since(start) [monotonic]", "monotonic", func() int64 { return int64(time.Since(start)) }},
		{"time.Now().UTC().UnixNano()", "realtime", func() int64 { return time.Now().UTC().UnixNano() }},
		{"runtime.nanotime via time.Since", "monotonic", func() int64 { return int64(time.Since(start)) }},
	}
	for _, r := range rows {
		fmt.Printf("  %-40s%-12s%12d ns\n", r.name, r.kind, measureResolution(r.read, 20))
	}

	now := time.Now()
	fmt.Println()
	fmt.Printf("  time.Now()          %v\n", now)
	fmt.Printf("  time.Now().UTC()    %v\n", now.UTC())
	fmt.Println("  ^ the 'm=+0.000...' suffix on the first one IS the monotonic reading.")
	fmt.Println("    It is not decoration. Its absence on the second line is section 3.")
}

// ------------------------------------------------- 2. one span, two clocks

// appClock is the application's own now(). Every service has one; most read the
// wall clock. The offset stands in for an NTP step -- we never touch the system
// clock, and lab/README.md explains why per-container skew is not possible here.
type appClock struct{ offset time.Duration }

func (c *appClock) now() time.Time       { return time.Now().Add(c.offset) }
func (c *appClock) step(d time.Duration) { c.offset += d }

func burn(d time.Duration) {
	end := time.Now().Add(d)
	for time.Now().Before(end) {
	}
}

func spanComparison(c *appClock) (wall, mono []float64) {
	// Fixed indices rather than a timer: a timer racing an 80ms loop is how you
	// get a run where the step lands between spans and the experiment silently
	// proves nothing -- which the README lists as a broken experiment, not a
	// wrong prediction.
	stepBackAt, stepFwdAt := spans/3, 2*spans/3
	for i := 0; i < spans; i++ {
		w0 := c.now()
		m0 := time.Now()
		burn(spanWorkU)
		if i == stepBackAt {
			c.step(stepBack)
		} else if i == stepFwdAt {
			c.step(-stepBack)
		}
		// w1.Sub(w0) would still use the monotonic readings both Times carry, so
		// the offset would be invisible. Going through UnixNano() is what an
		// application does when it stores or ships a timestamp -- and it is the
		// only honest way to model a service whose now() really is wall time.
		wall = append(wall, float64(c.now().UnixNano()-w0.UnixNano())/1e6)
		mono = append(mono, float64(time.Since(m0))/1e6)
	}
	return wall, mono
}

func pct(v []float64, q float64) float64 {
	s := append([]float64(nil), v...)
	sort.Float64s(s)
	i := int(q*float64(len(s))+0.5) - 1
	if i < 0 {
		i = 0
	}
	if i >= len(s) {
		i = len(s) - 1
	}
	return s[i]
}

func minMax(v []float64) (float64, float64) {
	lo, hi := v[0], v[0]
	for _, x := range v {
		if x < lo {
			lo = x
		}
		if x > hi {
			hi = x
		}
	}
	return lo, hi
}

func spanReport(wall, mono []float64) int {
	fmt.Println()
	fmt.Println("------------------------------------------------------------------------------")
	fmt.Printf("2. %d identical spans, timed twice, with a %v step and a %v step\n",
		spans, stepBack, -stepBack)
	fmt.Println("   landing INSIDE two of them")
	fmt.Println("------------------------------------------------------------------------------")
	fmt.Printf("  %-30s%10s%12s%14s%14s%10s\n", "clock", "p50", "p99", "max", "min", "negative")
	negatives := 0
	for _, row := range []struct {
		name string
		v    []float64
	}{{"wall (app now().UnixNano)", wall}, {"monotonic (time.Since)", mono}} {
		neg := 0
		for _, x := range row.v {
			if x < 0 {
				neg++
			}
		}
		if row.name[0] == 'w' {
			negatives = neg
		}
		lo, hi := minMax(row.v)
		fmt.Printf("  %-30s%10.3f%12.3f%14.1f%14.1f%10d\n",
			row.name, pct(row.v, 0.50), pct(row.v, 0.99), hi, lo, neg)
	}
	fmt.Println("  (milliseconds; 'negative' counts spans that finished before they started)")

	hot := 0
	for i, x := range wall {
		if x > wall[hot] {
			hot = i
		}
	}
	lo := hot - 19
	if lo < 0 {
		lo = 0
	}
	hi := hot + 21
	if hi > len(wall) {
		hi = len(wall)
	}
	wlo, whi := minMax(wall)
	fmt.Println()
	fmt.Printf("  Two samples out of %d were touched: %.0f ms and %.0f ms, against a p50\n",
		spans, wlo, whi)
	fmt.Printf("  of %.3f ms. Over all %d spans that is only the max -- one sample in %d\n",
		pct(wall, 0.50), spans, spans)
	fmt.Println("  cannot move a p99 by rank. But dashboards aggregate windows, not runs:")
	fmt.Printf("  over the %d spans around the step the wall-clock p99 is %.1f ms against\n",
		hi-lo, pct(wall[lo:hi], 0.99))
	fmt.Printf("  a monotonic p99 of %.3f ms. Only the clock differed.\n", pct(mono[lo:hi], 0.99))
	return negatives
}

// ------------------------------------------------------- 3. the Go footgun

func footguns() bool {
	fmt.Println()
	fmt.Println("------------------------------------------------------------------------------")
	fmt.Println("3. the footgun specific to this runtime: the stripped monotonic reading")
	fmt.Println("------------------------------------------------------------------------------")

	base := time.Now()
	var roundTripped time.Time
	if b, err := json.Marshal(base); err == nil {
		_ = json.Unmarshal(b, &roundTripped)
	}
	parsed, _ := time.Parse(time.RFC3339Nano, base.Format(time.RFC3339Nano))

	strippers := []struct {
		op string
		t  time.Time
	}{
		{"time.Now()            (untouched)", base},
		{"t.Round(0)", base.Round(0)},
		{"t.Truncate(time.Second)", base.Truncate(time.Second)},
		{"t.UTC()", base.UTC()},
		{"t.Local()", base.Local()},
		{"t.AddDate(0, 0, 0)", base.AddDate(0, 0, 0)},
		{"t.Add(time.Second)   (KEEPS it)", base.Add(time.Second)},
		{"JSON marshal + unmarshal", roundTripped},
		{"time.Parse(RFC3339Nano)", parsed},
	}
	fmt.Printf("  %-38s%-12s%s\n", "operation", "monotonic?", "printed form")
	for _, s := range strippers {
		// The only public way to ask is to print it: the "m=+..." suffix appears
		// exactly when a monotonic reading is present. There is no API for this,
		// which is a large part of why the bug survives code review.
		text := s.t.String()
		has := "STRIPPED"
		if idx := indexOf(text, " m=+"); idx >= 0 {
			has = "kept"
			text = text[:idx] + " " + text[idx+1:]
		}
		if len(text) > 34 {
			text = text[:34]
		}
		fmt.Printf("  %-38s%-12s%s\n", s.op, has, text)
	}

	// Now the bug itself. Same span, two ways: one keeps its monotonic readings,
	// the other round-trips the start through .UTC() the way any code that logs,
	// serialises or stores a timestamp would.
	fmt.Println()
	start := time.Now()
	stripped := start.UTC()
	burn(2 * time.Millisecond)
	end := time.Now()

	good := end.Sub(start)
	bad := end.Sub(stripped)
	fmt.Printf("  end.Sub(start)          %v      [monotonic path -- correct]\n", good)
	fmt.Printf("  end.Sub(start.UTC())    %v      [wall path -- same instants!]\n", bad)
	fmt.Println("  Both operands describe the same two moments. One call to .UTC() moved")
	fmt.Println("  the subtraction onto the wall clock, and nothing anywhere said so.")

	// And with a wall clock that has stepped, the wall path goes negative while
	// the monotonic path is untouched. We cannot step the real clock, so we do
	// what a stepped clock does to the arithmetic: subtract Times whose wall
	// readings disagree with their ordering.
	fmt.Println()
	c := &appClock{}
	w0 := c.now().UTC() // stripped: only the wall reading survives
	burn(2 * time.Millisecond)
	c.step(stepBack) // NTP decides we are ahead and jumps back
	w1 := c.now().UTC()
	fmt.Printf("  with a %v step inside the span:\n", stepBack)
	fmt.Printf("    w1.Sub(w0) on stripped Times   %v   <-- NEGATIVE DURATION\n", w1.Sub(w0))
	m0 := time.Now()
	burn(2 * time.Millisecond)
	fmt.Printf("    time.Since(m0), monotonic      %v\n", time.Since(m0))
	fmt.Println("  A negative time.Duration is a legal value in Go. It will divide, average")
	fmt.Println("  and export to your metrics pipeline without complaint.")
	return w1.Sub(w0) < 0
}

func indexOf(s, sub string) int {
	for i := 0; i+len(sub) <= len(s); i++ {
		if s[i:i+len(sub)] == sub {
			return i
		}
	}
	return -1
}

func main() {
	fmt.Println("==============================================================================")
	fmt.Println("Layer 4 Topic 3 -- Go clock audit")
	fmt.Println("==============================================================================")
	fmt.Printf("  %s on %s/%s\n\n", runtime.Version(), runtime.GOOS, runtime.GOARCH)

	inventory()
	c := &appClock{}
	wall, mono := spanComparison(c)
	negatives := spanReport(wall, mono)
	reproduced := footguns()

	fmt.Println()
	fmt.Println("------------------------------------------------------------------------------")
	fmt.Println("4. one line for the record table in the README")
	fmt.Println("------------------------------------------------------------------------------")
	start := time.Now()
	res := measureResolution(func() int64 { return int64(time.Since(start)) }, 20)
	verdict := "NO -- investigate"
	if reproduced {
		verdict = "yes"
	}
	plural := "s"
	if negatives == 1 {
		plural = ""
	}
	fmt.Printf("  | Go | time.Time's hidden monotonic reading | %d ns | %s (%d negative wall-clock span%s) |\n",
		res, verdict, negatives, plural)
	fmt.Println()
	fmt.Println("  The table in the README stays blank until you fill it in. This line is")
	fmt.Println("  the measurement, not the answer -- copy it across yourself.")
}

File: 04-distributed/03-clocks-lie/nodejs/clock_audit.js
// Layer 4 Topic 3 (Part A) -- Node's clocks, audited rather than assumed.
//
// WHAT THIS DEMONSTRATES: four things, in order.
//   1. the clock inventory: Date.now() is the wall clock and is settable;
//      performance.now() is monotonic sub-millisecond; process.hrtime.bigint()
//      is monotonic nanoseconds. Resolution is measured here, not quoted.
//   2. one span timed twice -- through the application's own now(), which reads
//      the wall clock, and through process.hrtime.bigint() -- while an
//      NTP-style step is applied inside two of the spans.
//   3. Node's specific subtlety: performance.timeOrigin is a WALL-CLOCK anchor
//      captured at process start, so `timeOrigin + performance.now()` inherits
//      every wall-clock hazard while performance.now() alone does not. Mixing
//      the two in one calculation is how a monotonic measurement gets
//      contaminated, and this program does it on purpose so you can see it.
//   4. the summary line for the README's record table.
//
// WHAT TO LOOK FOR IN THE OUTPUT: the NEGATIVE column in section 2, and in
// section 3 the fact that the `timeOrigin + performance.now()` span reports the
// same corrupted duration as the raw wall clock -- despite being built out of a
// monotonic reading.
//
//   node nodejs/clock_audit.js

'use strict';

const os = require('node:os');
const { performance } = require('node:perf_hooks');

const STEP_MS = -40000;   // an NTP correction, applied backwards, mid-run
const SPANS = 400;
const SPAN_WORK_US = 200;

// ------------------------------------------------------------- 1. inventory

/** Smallest non-zero delta this clock will report. Measured, not documented --
 *  a clock can advertise nanoseconds and tick in microseconds. */
function measureResolutionMs(read, trials = 20) {
  let smallest = Infinity;
  for (let t = 0; t < trials; t++) {
    const a = read();
    for (;;) {
      const b = read();
      if (b !== a) {
        const d = Number(b > a ? b - a : a - b);
        if (d < smallest) smallest = d;
        break;
      }
    }
  }
  return smallest;
}

function inventory() {
  console.log('-'.repeat(78));
  console.log('1. the clocks Node offers, and what each one is for');
  console.log('-'.repeat(78));
  console.log(`  ${'expression'.padEnd(34)}${'kind'.padEnd(12)}${'settable'.padEnd(10)}measured resolution`);

  const rows = [
    ['Date.now()', 'realtime', 'YES', () => Date.now(), 1e6],
    ['performance.now()', 'monotonic', 'no', () => performance.now(), 1e6],
    ['process.hrtime.bigint()', 'monotonic', 'no', () => process.hrtime.bigint(), 1],
    ['performance.timeOrigin + now()', 'realtime', 'YES',
      () => performance.timeOrigin + performance.now(), 1e6],
  ];
  for (const [name, kind, settable, read, toNs] of rows) {
    const res = measureResolutionMs(read) * toNs;
    console.log(`  ${name.padEnd(34)}${kind.padEnd(12)}${settable.padEnd(10)}${String(Math.round(res)).padStart(12)} ns`);
  }
  console.log();
  console.log(`  performance.timeOrigin = ${performance.timeOrigin.toFixed(3)} ms`);
  console.log('  ^ a wall-clock reading captured once, at process start. Everything you');
  console.log('    add it to becomes a wall-clock value, however monotonic it started out.');
}

// ------------------------------------------------- 2. one span, two clocks

/** The application's own now(). Every service has one; most read the wall clock.
 *  The offset stands in for an NTP step -- we never touch the system clock, and
 *  lab/README.md explains why per-container skew is not even possible here. */
class AppClock {
  constructor() { this.offsetMs = 0; }
  now() { return Date.now() + this.offsetMs; }
  step(ms) { this.offsetMs += ms; }
}

function burn(micros) {
  const end = process.hrtime.bigint() + BigInt(micros) * 1000n;
  while (process.hrtime.bigint() < end) { /* busy: we are timing, not sleeping */ }
}

function spanComparison(clock) {
  const wall = [];
  const mono = [];
  const contaminated = [];
  // Fixed indices, not a timer. A timer racing an 80ms loop is how you get a run
  // where the step lands between spans and the experiment silently proves
  // nothing -- which the README lists as a broken experiment, not a wrong
  // prediction. Determinism about *when* buys realism about *what*.
  const stepBackAt = Math.floor(SPANS / 3);
  const stepFwdAt = Math.floor((2 * SPANS) / 3);

  for (let i = 0; i < SPANS; i++) {
    const w0 = clock.now();
    const m0 = process.hrtime.bigint();
    const c0 = performance.timeOrigin + performance.now() + clock.offsetMs;
    burn(SPAN_WORK_US);
    if (i === stepBackAt) clock.step(STEP_MS);
    else if (i === stepFwdAt) clock.step(-STEP_MS);
    const w1 = clock.now();
    const m1 = process.hrtime.bigint();
    const c1 = performance.timeOrigin + performance.now() + clock.offsetMs;
    wall.push(w1 - w0);
    mono.push(Number(m1 - m0) / 1e6);
    contaminated.push(c1 - c0);
  }
  return { wall, mono, contaminated };
}

function pct(values, q) {
  const s = [...values].sort((a, b) => a - b);
  return s[Math.min(s.length - 1, Math.max(0, Math.round(q * s.length + 0.5) - 1))];
}

function spanReport({ wall, mono, contaminated }) {
  console.log();
  console.log('-'.repeat(78));
  console.log(`2. ${SPANS} identical spans, timed three ways, with a ${STEP_MS / 1000}s step`);
  console.log(`   and a ${-STEP_MS / 1000}s step landing INSIDE two of them`);
  console.log('-'.repeat(78));
  const head = `  ${'clock'.padEnd(30)}${'p50'.padStart(10)}${'p99'.padStart(12)}${'max'.padStart(14)}${'min'.padStart(14)}${'negative'.padStart(10)}`;
  console.log(head);
  let negatives = 0;
  const rows = [
    ['wall (app now())', wall],
    ['monotonic (hrtime.bigint)', mono],
    ['timeOrigin + performance.now()', contaminated],
  ];
  for (const [name, v] of rows) {
    const neg = v.filter((x) => x < 0).length;
    if (name.startsWith('wall')) negatives = neg;
    console.log(`  ${name.padEnd(30)}${pct(v, 0.5).toFixed(3).padStart(10)}${pct(v, 0.99).toFixed(3).padStart(12)}` +
      `${Math.max(...v).toFixed(1).padStart(14)}${Math.min(...v).toFixed(1).padStart(14)}${String(neg).padStart(10)}`);
  }
  console.log("  (milliseconds; 'negative' counts spans that finished before they started)");
  console.log();
  const hot = wall.indexOf(Math.max(...wall));
  const lo = Math.max(0, hot - 19);
  const win = wall.slice(lo, hot + 21);
  const winMono = mono.slice(lo, hot + 21);
  console.log(`  Two samples out of ${SPANS} were touched: ${Math.min(...wall).toFixed(0)} ms and ${Math.max(...wall).toFixed(0)} ms,`);
  console.log(`  against a p50 of ${pct(wall, 0.5).toFixed(3)} ms. Over all ${SPANS} spans that is only the max --`);
  console.log(`  one sample in ${SPANS} cannot move a p99 by rank. But dashboards aggregate`);
  console.log(`  windows, not runs: over the ${win.length} spans around the step the wall-clock`);
  console.log(`  p99 is ${pct(win, 0.99).toFixed(1)} ms against a monotonic p99 of ${pct(winMono, 0.99).toFixed(3)} ms.`);
  console.log('  Same workload, same machine, same instant. Only the clock differed.');
  console.log();
  console.log(`  Read the wall row's p50 too: ${pct(wall, 0.5).toFixed(3)} ms for work the monotonic`);
  console.log(`  row measures at ${pct(mono, 0.5).toFixed(3)} ms. Date.now() has millisecond resolution, so`);
  console.log('  it cannot see this span AT ALL -- it reports 0 or 1 and nothing between.');
  console.log('  That is a second, quieter reason not to time spans with the wall clock.');
  return negatives;
}

// ----------------------------------------------------- 3. the Node footgun

function footguns() {
  console.log();
  console.log('-'.repeat(78));
  console.log('3. the footgun specific to this runtime: timeOrigin contamination');
  console.log('-'.repeat(78));

  // performance.now() is monotonic. Adding timeOrigin to it makes a wall-clock
  // estimate -- and the row above proves it, because the contaminated column
  // tracked the wall clock through the step while the monotonic column did not.
  const t0mono = performance.now();
  const t0wall = performance.timeOrigin + performance.now();
  burn(1000);
  const t1mono = performance.now();
  const t1wall = performance.timeOrigin + performance.now();
  console.log(`  performance.now() span                 ${(t1mono - t0mono).toFixed(4)} ms   [monotonic]`);
  console.log(`  (timeOrigin + performance.now()) span  ${(t1wall - t0wall).toFixed(4)} ms   [wall clock]`);
  console.log('  The two agree here because nothing stepped. They are still different');
  console.log('  kinds of number, and only one of them survives an NTP correction.');
  console.log();

  // The other half of the subtlety: Date.now() and timeOrigin+now() are both
  // wall-clock, so they agree -- but they are anchored at different instants,
  // and the drift between them is real and grows.
  const drift = (performance.timeOrigin + performance.now()) - Date.now();
  console.log(`  (timeOrigin + performance.now()) - Date.now() = ${drift.toFixed(3)} ms`);
  console.log('  ^ both are "the wall clock". They disagree because timeOrigin was');
  console.log('    sampled once, at startup, and has been accumulating drift ever since.');
  console.log('    In a process that has been up for a week this is not a rounding error.');
  console.log();

  console.log(`  process.uptime()      ${process.uptime().toFixed(3)} s   [monotonic-ish, since process start]`);
  console.log(`  os.uptime()           ${os.uptime()} s   [since boot]`);
  console.log(`  platform              ${process.platform} ${os.release()} ${os.arch()}`);
  console.log('  Node exposes no CLOCK_MONOTONIC_RAW, no CLOCK_BOOTTIME and no way to');
  console.log('  ask which OS clock backs performance.now(). If you need that, you are');
  console.log('  in the wrong runtime for the question -- see cpp/clock_audit.cpp.');
  return true;
}

function main() {
  console.log('='.repeat(78));
  console.log('Layer 4 Topic 3 -- Node.js clock audit');
  console.log('='.repeat(78));
  console.log(`  Node ${process.version} on ${process.platform} ${os.arch()}`);
  console.log();
  inventory();
  const negatives = spanReport(spanComparison(new AppClock()));
  const reproduced = footguns();

  console.log();
  console.log('-'.repeat(78));
  console.log("4. one line for the record table in the README");
  console.log('-'.repeat(78));
  const res = measureResolutionMs(() => process.hrtime.bigint());
  console.log(`  | Node.js | process.hrtime.bigint() | ${res} ns | ` +
    `${reproduced ? 'yes' : 'NO -- investigate'} ` +
    `(${negatives} negative wall-clock span${negatives === 1 ? '' : 's'}) |`);
  console.log();
  console.log('  The table in the README stays blank until you fill it in. This line is');
  console.log('  the measurement, not the answer -- copy it across yourself.');
}

main();
